cmd/snap-confine/seccomp-support.c

```diff
@@ -17,869 +17,182 @@
 
-	f = fopen(profile_path, "r");
-	if (f == NULL) {
-		fprintf(stderr, "Can not open %s (%s)\n", profile_path,
-			strerror(errno));
-		die("aborting");
-	}
-	// Note, preprocess_filter() die()s on error
-	preprocess_filter(f, &pre);
-
-	if (pre.unrestricted) {
-		seccomp_release(ctx);
-		ctx = NULL;
-		goto out;
-	}
-	// FIXME: right now complain mode is the equivalent to unrestricted.
-	// We'll want to change this once we seccomp logging is in order.
-	if (pre.complain) {
-		seccomp_release(ctx);
-		ctx = NULL;
-		goto out;
-	}
-
-	char buf[SC_MAX_LINE_LENGTH];
-	while (fgets(buf, sizeof(buf), f) != NULL) {
-		lineno++;
-
-		// skip policy-irrelevant lines
-		if (validate_and_trim_line(buf, sizeof(buf), lineno) == 0)
-			continue;
-
-		char *buf_copy = strdup(buf);
-		if (buf_copy == NULL)
-			die("Out of memory");
-
-		int pr_rc = parse_line(buf_copy, &sargs);
-		free(buf_copy);
-		if (pr_rc != PARSE_OK) {
-			// as this is a syscall whitelist an invalid syscall
-			// is ok and the error can be ignored
-			if (pr_rc == PARSE_INVALID_SYSCALL)
-				continue;
-			die("could not parse line");
-		}
-
-		rc = seccomp_rule_add_exact_array(ctx, SCMP_ACT_ALLOW,
-						  sargs.syscall_nr,
-						  sargs.length, sargs.arg_cmp);
-		if (rc != 0) {
-			rc = seccomp_rule_add_array(ctx, SCMP_ACT_ALLOW,
-						    sargs.syscall_nr,
-						    sargs.length,
-						    sargs.arg_cmp);
-			if (rc != 0) {
-				fprintf(stderr,
-					"seccomp_rule_add_array failed with %i for '%s'\n",
-					rc, buf);
-				errno = 0;
-				die("aborting");
-			}
-		}
-	}
-
- out:
-	if (f != NULL) {
-		if (fclose(f) != 0)
-			die("could not close seccomp file");
-	}
-	sc_map_destroy();
-	return ctx;
-}
-
-void sc_load_seccomp_context(scmp_filter_ctx ctx)
-{
-	int rc;
+	// validate '/' down to profile_path are root-owned and not
+	// 'other' writable to avoid possibility of privilege
+	// escalation via bpf program load when paths are incorrectly
+	// set on the system.
+	validate_bpfpath_is_safe(profile_path);
+
+	// sanity check size/owner
+	struct stat stat_buf;
+	if (stat(profile_path, &stat_buf) < 0)
+		die("cannot stat %s", profile_path);
+	if (stat_buf.st_size > MAX_BPF_SIZE)
+		die("profile %s is too big", profile_path);
+
+	// paranoid, but only helps a bit
+	if (stat_buf.st_uid != 0 || stat_buf.st_gid != 0)
+		die("cannot use %s: must be root:root owned", profile_path);
+
+	// load bpf
+	unsigned char bpf[MAX_BPF_SIZE + 1];	// acount for EOF
+	FILE *fp = fopen(profile_path, "rb");
+	if (fp == NULL)
+		die("cannot read %s", profile_path);
+	// set 'size' to 1 to get bytes transferred
+	size_t num_read = fread(bpf, 1, sizeof(bpf), fp);
+	if (ferror(fp) != 0)
+		die("cannot fread() %s", profile_path);
+	else if (feof(fp) == 0)
+		die("profile %s exceeds %zu bytes", profile_path, sizeof(bpf));
+	fclose(fp);
+	debug("read %zu bytes from %s", num_read, profile_path);
+
+	// validate bpf, it will die() if things look wrong
+	validate_bpf(bpf, num_read);
+
+	// raise privs
```

Comment:
Can you get rid of this comment? It ended up before the `getresuid()` call and then after the `getresuid()` there is a comment about raising.